Maze.cpp

```diff
@@ -17,17 +18,101 @@ const CellCoordinate Maze::CELL_FINISH =
 void Maze::reset() {
     for (int y = 0; y < NODE_ROWS; y++) {
         for (int x = 0; x < NODE_COLS; x++) {
-            maze[y][x] = Node(NodeCoordinate(x, y));
+            NodeCoordinate pos = NodeCoordinate(x, y);
+            maze[y][x] = Node(pos);
 
             // Initialize border
-            if (x == 0 || y == 0 || y == NODE_ROWS - 1 || x == NODE_COLS - 1)
-                setWall(NodeCoordinate(x, y));
+            if (isBorder(pos) || pos.isPost())
+                setWall(pos);
         }
     }
-
+    setWall(NodeCoordinate(NODE_COLS / 2, NODE_ROWS / 2), false);
     path.clear();
 }
 
+bool Maze::isBorder(NodeCoordinate pos) {
+    return pos.x == 0 || pos.y == 0 || pos.y == NODE_ROWS - 1 ||
+           pos.x == NODE_COLS - 1;
+}
+
+bool Maze::isBorder(Node node) {
+    return isBorder(node.pos);
+}
+
+bool Maze::withinBounds(NodeCoordinate pos) {
+    return pos.x >= 0 && pos.x < NODE_COLS && pos.y >= 0 && pos.y < NODE_ROWS;
+}
+
+bool Maze::withinBounds(Node node) {
+    return withinBounds(node.pos);
+}
+
+void Maze::generate(int seed) {
+    reset();
+    srand(seed);
+    vector<CellCoordinate> backtrackStack;
+    set<CellCoordinate> unvisited;
+    backtrackStack.reserve(CELL_COLS * CELL_ROWS);
+
+    for (int i = 0; i < CELL_ROWS; i++) {
+        for (int j = 0; j < CELL_COLS; j++) {
+            unvisited.insert(CellCoordinate(j, i));
+            setWall(CellCoordinate(j, i).toNode() + N);
+            setWall(CellCoordinate(j, i).toNode() + E);
+            setWall(CellCoordinate(j, i).toNode() + W);
+            setWall(CellCoordinate(j, i).toNode() + S);
+        }
+    }
+    CellCoordinate currentCell = CellCoordinate(0, 0);
+    unvisited.erase(currentCell);
+    vector<CellCoordinate> neighbors;
+    neighbors.reserve(4);
+    while (unvisited.size() > 0) {
+        if (withinBounds(currentCell + N) &&
+            unvisited.find(currentCell + N) != unvisited.end()) {
+            neighbors.push_back(currentCell + N);
+        }
+        if (withinBounds(currentCell + E) &&
+            unvisited.find(currentCell + E) != unvisited.end()) {
+            neighbors.push_back(currentCell + E);
+        }
+        if (withinBounds(currentCell + W) &&
+            unvisited.find(currentCell + W) != unvisited.end()) {
+            neighbors.push_back(currentCell + W);
+        }
```

Comment:
A lot of repetition of code here, try a for-loop. Look at `findPath` if you need a reference.